metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -1500,12 +1501,23 @@ public boolean addPartition(Partition part) throws InvalidObjectException,
   @Override
   public Partition getPartition(String dbName, String tableName,
       List<String> part_vals) throws NoSuchObjectException, MetaException {
-    openTransaction();
-    Partition part = convertToPart(getMPartition(dbName, tableName, part_vals));
-    commitTransaction();
-    if(part == null) {
-      throw new NoSuchObjectException("partition values="
+    boolean committed = false;
+    Partition part = null;
```

Comment:
Do we even need a openTransaction here? getMPartition() opens and closes the transaction in a try-finally block, so I think all we need to do here is to remove the openTransaction() and commitTransaction() from this method. The NucleusObjectNotFoundException is a runtime exception from datanucleus which we don't catch (as can be seen in other methods), so I would suggest you to not catch it here. Basically, all this patch would require is to remove the openTransaction/commitTransaction() pair and it should work.

Reply:
In the master branch, the code is different. To implement [HIVE-19915](https://issues.apache.org/jira/browse/HIVE-19915) we need to get the table and the partition together in one transaction.

Reply:
I suggest we make the changes into master first and then backport it to hive-1, hive-2 branches.

Reply:
@vihangk1 Thanks for the review, that makes sense to me. Change updated.

Reply:
Hi @kmanamcheri , the implementation has changed so much on master, actually I am not sure if this is still a problem on master.

Reply:
Actually, I think the patch applies on the master branch as well. In master branch we can't get rid of the transaction since there is a getTable operation as well. So we can wrap the code in a try { } finally {} block. I am not sure if branch-1 HiveQA job is still working (been long time since we had a branch-1 patch). If you could provide a attach the patch for master branch, HiveQA will trigger tests on it. While committing, I can manually port it to branch-2 and branch-1. Also, having a unit-test would be great too!

Reply:
@guangyy as vihang was stating, your patch actually applies on master as well. The files have moved around a bit and so it might not apply cleanly!

The getPartition(..) call is [here](https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L2443) in the master branch. If you look at that call, you can see that there is no try..catch block. This appears to be a problem on master. Not sure if it shows up in any tests though!

+1 to unit tests